### PR TITLE
Make Api `no_std` compatible

### LINKED ISSUE #279: Be fully `no_std` compatible (except for ws rpc)
Would include hosting a ws client in no-std. To my knowledge, this is not yet available as a library.

### PR DESCRIPTION
**Pull request by haerdib**
🌈  Featuring added `no_std` compatibility for the full `Api` included traits and the implementations (except for the rpc clients)
- removes `sp-std` dependency. Whenever possible, depend on `core` or `alloc` directly. The `std` version itself points to them as well. No need for an extra library to do that,
- Introduces a lot of substrate types in `primitives/serde_impls` to implement serde also in `no_std`. Wherever sensible, the conversion from substrate type and redefined type has been implemented with `From`. So the newly self-implemented types should not pose a big problem for the user.  Another good thing: If substrate updates the types, this will be noticed quite fast via the unit tests.
- Removed `MaybeSerializeDeserialize` because this only enforces (De)Serialization only in `std` mode.
- Reimplemented `NumberOrHex` because  `sp-rpc` is not `no_std` compatible. 
- Removed some unused error types in `src/api/error.rs`

closes #279

### CODE REVIEW

**Review by clangenb — COMMENTED** [review]
Very cool we were not so far away after all! :D I only have a comment about documentation.

File: primitives/src/pallet_traits/pallet_assets_config.rs [hunk 1]
```diff
@@ -25,11 +26,12 @@ pub trait AssetsConfig: crate::FrameSystemConfig {
 	type Balance: AtLeast32BitUnsigned
 		+ Default
 		+ Copy
-		+ MaybeSerializeDeserialize
+		+ Serialize
+		+ DeserializeOwned
```
[the remark this answers is not in the document]
Comment (haerdib) [reply primitives/src/pallet_traits/pallet_assets_config.rs:30]: Yeah, I agree. I also added two issues in that regard: https://github.com/scs/substrate-api-client/issues/393 & https://github.com/scs/substrate-api-client/issues/391

But a comment here would not hurt.

**Review by Niederb — APPROVED** [review]
LGTM and a nice improvement. The type duplication from substrate is a bit unfortunate, let's hope they agree to provide serialization also in no_std environments.

File: primitives/src/types.rs [hunk 2]
```diff
@@ -133,25 +125,26 @@ pub struct RuntimeDispatchInfo<Balance, Weight = sp_weights::OldWeight> {
 	///
 	/// This does not include a tip or anything else that
 	/// depends on the signature (i.e. depends on a `SignedExtension`).
-	#[cfg_attr(feature = "std", serde(with = "serde_balance"))]
+	//#[cfg_attr(feature = "std", serde(with = "serde_balance"))]
```
[the remark this answers is not in the document]
Comment (haerdib) [reply primitives/src/types.rs]: Yup, you're right, thanks a lot!

**Review by clangenb — APPROVED** [review]
Beautiful, thanks for the extra documentation!

**Pull request merged**